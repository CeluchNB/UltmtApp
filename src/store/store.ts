import accountReducer from './reducers/features/account/accountReducer'
import { configureStore } from '@reduxjs/toolkit'
import liveGameReducer from './reducers/features/game/liveGameReducer'
import livePointReducer from './reducers/features/point/livePointReducer'
import managedTeamReducer from './reducers/features/team/managedTeamReducer'

const store = configureStore({
    reducer: {
        account: accountReducer,
        managedTeam: managedTeamReducer,
        liveGame: liveGameReducer,
        livePoint: livePointReducer,
    },
})

export type RootState = ReturnType<typeof store.getState>
export default store
