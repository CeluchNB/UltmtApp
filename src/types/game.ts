import { Tournament } from './tournament'
import { DisplayTeam, GuestTeam } from './team'
import { DisplayUser, GuestUser } from './user'
export interface DisplayGame {
    opponent: DisplayTeam
    teamScore: number
    opponentScore: number
}

export interface CreateGame {
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
}

export interface Game {
    _id: string
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
    teamOneScore: number
    teamTwoScore: number
    teamOneActive: boolean
    teamTwoActive: boolean
    teamOnePlayers: DisplayUser[]
    teamTwoPlayers: DisplayUser[]
    resolveCode: string
    points: string[]
}

export interface SearchParams {
    q?: string
    live?: boolean
    after?: Date
    before?: Date
}
