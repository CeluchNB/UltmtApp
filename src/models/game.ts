import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { Tournament } from '../types/tournament'
import { CreateGame, Game, GameStatus, PointStats } from '../types/game'
import { DisplayTeam, GuestTeam } from '../types/team'

export class GameSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Game',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            creator: 'DisplayUser',
            teamOne: 'DisplayTeam',
            teamTwo: 'GuestTeam',
            teamTwoDefined: 'bool',
            scoreLimit: 'int',
            halfScore: 'int',
            startTime: 'date',
            softcapMins: 'int',
            hardcapMins: 'int',
            playersPerPoint: 'int',
            timeoutPerHalf: 'int',
            floaterTimeout: 'bool',
            tournament: 'Tournament?',
            teamOneScore: 'int',
            teamTwoScore: 'int',
            teamOnePlayers: 'DisplayUser[]',
            teamTwoPlayers: 'DisplayUser[]',
            resolveCode: 'string',
            statsPoints: 'PointStats[]',
            offline: 'bool',
            teamOneStatus: 'string',
            teamTwoStatus: 'string',
        },
    }

    _id: string
    creator: DisplayUser
    teamOne: DisplayTeam
    teamTwo: GuestTeam
    teamTwoDefined: boolean
    scoreLimit: number
    halfScore: number
    startTime: Date
    softcapMins: number
    hardcapMins: number
    playersPerPoint: number
    timeoutPerHalf: number
    floaterTimeout: boolean
    tournament?: Tournament
    teamOneScore: number
    teamTwoScore: number
    teamOnePlayers: DisplayUser[]
    teamTwoPlayers: DisplayUser[]
    resolveCode: string
    statsPoints: PointStats[]
    offline: boolean
    teamOneStatus: GameStatus
    teamTwoStatus: GameStatus

    constructor(
        game: Game,
        offline: boolean = false,
        statsPoints: PointStats[] = [],
    ) {
        this._id = game._id
        this.creator = game.creator
        this.teamOne = game.teamOne
        this.teamTwo = game.teamTwo
        this.teamTwoDefined = game.teamTwoDefined
        this.scoreLimit = game.scoreLimit
        this.halfScore = game.halfScore
        this.startTime = game.startTime
        this.softcapMins = game.softcapMins
        this.hardcapMins = game.hardcapMins
        this.playersPerPoint = game.playersPerPoint
        this.timeoutPerHalf = game.timeoutPerHalf
        this.floaterTimeout = game.floaterTimeout
        this.tournament = game.tournament
        this.teamOneScore = game.teamOneScore
        this.teamTwoScore = game.teamTwoScore
        this.teamOnePlayers = game.teamOnePlayers
        this.teamTwoPlayers = game.teamTwoPlayers
        this.resolveCode = game.resolveCode || ''
        this.statsPoints = statsPoints
        this.offline = offline
        this.teamOneStatus = game.teamOneStatus
        this.teamTwoStatus = game.teamTwoStatus
    }

    static createOfflineGame(game: CreateGame, teamOnePlayers: DisplayUser[]) {
        return {
            _id: new Realm.BSON.ObjectID().toHexString(),
            creator: game.creator,
            teamOne: game.teamOne,
            teamTwo: game.teamTwo,
            teamTwoDefined: game.teamTwoDefined,
            scoreLimit: game.scoreLimit,
            halfScore: game.halfScore,
            startTime: new Date(game.startTime),
            softcapMins: game.softcapMins,
            hardcapMins: game.hardcapMins,
            playersPerPoint: game.playersPerPoint,
            timeoutPerHalf: game.timeoutPerHalf,
            floaterTimeout: game.floaterTimeout,
            tournament: game.tournament,
            teamOneScore: 0,
            teamTwoScore: 0,
            teamOneActive: true,
            teamTwoActive: false,
            teamOnePlayers: teamOnePlayers,
            teamTwoPlayers: [],
            resolveCode: '',
            statsPoints: [],
            offline: true,
        }
    }
}
export const PointStatsSchema: Realm.ObjectSchema = {
    name: 'PointStats',
    embedded: true,
    properties: {
        _id: 'string',
        pointStats: 'PlayerPointStats[]',
    },
}

export const PlayerPointStatsSchema: Realm.ObjectSchema = {
    name: 'PlayerPointStats',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string?',
        pointsPlayed: 'int',
        goals: 'int',
        assists: 'int',
        turnovers: 'int',
        blocks: 'int',
    },
}
