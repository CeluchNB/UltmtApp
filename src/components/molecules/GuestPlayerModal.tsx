import { AppDispatch } from '../../store/store'
import BaseModal from '../atoms/BaseModal'
import { GuestUser } from '../../types/user'
import PrimaryButton from '../atoms/PrimaryButton'
import React from 'react'
import UserInput from '../atoms/UserInput'
import { addGuestPlayer } from '../../store/reducers/features/game/liveGameReducer'
import { getFormFieldRules } from '../../utils/form-utils'
import { useTheme } from '../../hooks'
import { Controller, useForm } from 'react-hook-form'
import { StyleSheet, Text, View } from 'react-native'
import {
    resetGuestPlayerStatus,
    selectGuestPlayerError,
    selectGuestPlayerStatus,
} from '../../store/reducers/features/game/liveGameReducer'
import { useDispatch, useSelector } from 'react-redux'

interface GuestPlayerModalProps {
    visible: boolean
    onClose: () => void
}

const GuestPlayerModal: React.FC<GuestPlayerModalProps> = ({
    visible,
    onClose,
}) => {
    const {
        theme: { colors, size },
    } = useTheme()
    const status = useSelector(selectGuestPlayerStatus)
    const error = useSelector(selectGuestPlayerError)
    const dispatch = useDispatch<AppDispatch>()

    const {
        control,
        handleSubmit,
        formState: { errors },
        reset,
    } = useForm({
        defaultValues: { firstName: '', lastName: '' },
    })

    const onSubmitPlayer = async (player: GuestUser) => {
        dispatch(addGuestPlayer(player))
    }

    React.useEffect(() => {
        if (status === 'success') {
            reset()
            dispatch(resetGuestPlayerStatus())
            onClose()
        }
    }, [status, onClose, reset, dispatch])

    const styles = StyleSheet.create({
        modalContainer: {
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            marginTop: 30,
        },
        modalView: {
            margin: 20,
            backgroundColor: colors.darkGray,
            borderRadius: 10,
            padding: 25,
            alignItems: 'center',
            shadowColor: colors.darkPrimary,
            shadowOffset: {
                width: 5,
                height: 5,
            },
            shadowOpacity: 0.25,
            shadowRadius: 4,
            elevation: 5,
        },
        title: {
            fontSize: size.fontTwenty,
            color: colors.textSecondary,
        },
        inputContainer: {
            flexDirection: 'row',
            display: 'flex',
        },
        input: {
            width: '90%',
        },
        errorText: {
            color: colors.error,
            fontSize: size.fontFifteen,
        },
        button: {
            marginTop: 10,
        },
    })

    return (
        <BaseModal visible={visible} onClose={onClose}>
            <Text style={styles.title}>Add Guest Player</Text>
            <View style={styles.inputContainer}>
                <Controller
                    control={control}
                    name="firstName"
                    rules={getFormFieldRules('First name', true, 1, 30)}
                    render={({ field: { onChange, value } }) => {
                        return (
                            <UserInput
                                style={styles.input}
                                placeholder="First Name"
                                onChangeText={onChange}
                                value={value}
                            />
                        )
                    }}
                />
            </View>
            {errors.firstName && (
                <Text style={styles.errorText}>{errors.firstName.message}</Text>
            )}
            <View style={styles.inputContainer}>
                <Controller
                    control={control}
                    name="lastName"
                    rules={getFormFieldRules('Last name', true, 1, 30)}
                    render={({ field: { onChange, value } }) => {
                        return (
                            <UserInput
                                style={styles.input}
                                placeholder="Last Name"
                                onChangeText={onChange}
                                value={value}
                            />
                        )
                    }}
                />
            </View>
            {errors.lastName && (
                <Text style={styles.errorText}>{errors.lastName.message}</Text>
            )}
            {error && <Text style={styles.errorText}>{error}</Text>}
            <PrimaryButton
                style={styles.button}
                text="add"
                onPress={async () => {
                    handleSubmit(onSubmitPlayer)()
                }}
                loading={status === 'loading'}
            />
        </BaseModal>
    )
}

export default GuestPlayerModal
