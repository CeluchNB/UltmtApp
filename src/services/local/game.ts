import * as Constants from '../../utils/constants'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { DisplayUser } from '../../types/user'
import { getRealm } from '../../models/realm'
import { throwApiError } from '../../utils/service-utils'
import { ActionSchema, GameSchema, PointSchema } from '../../models'
import { CreateGame, Game } from '../../types/game'

const parseGame = (schema: GameSchema): Game & { offline: boolean } => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            creator: schema.creator,
            teamOne: schema.teamOne,
            teamTwo: schema.teamTwo,
            teamTwoDefined: schema.teamTwoDefined,
            scoreLimit: schema.scoreLimit,
            halfScore: schema.halfScore,
            startTime: schema.startTime,
            softcapMins: schema.softcapMins,
            hardcapMins: schema.hardcapMins,
            playersPerPoint: schema.playersPerPoint,
            timeoutPerHalf: schema.timeoutPerHalf,
            floaterTimeout: schema.floaterTimeout,
            tournament: schema.tournament,
            teamOneScore: schema.teamOneScore,
            teamTwoScore: schema.teamTwoScore,
            teamOneActive: schema.teamOneActive,
            teamTwoActive: schema.teamTwoActive,
            teamOnePlayers: schema.teamOnePlayers,
            teamTwoPlayers: schema.teamTwoPlayers,
            resolveCode: schema.resolveCode,
            points: schema.points,
            offline: schema.offline,
        }),
    )
}

/**
 * Method to determine if currently active game is offline
 * @returns boolean indicating current game is offline or not
 */
export const activeGameOffline = async (): Promise<boolean> => {
    return (await AsyncStorage.getItem('active_game_offline')) === 'true'
}

/**
 * Method to get the ObjectId of the current active game
 * @returns current active game id
 */
export const activeGameId = async (): Promise<string> => {
    return (await AsyncStorage.getItem('active_game_id')) || ''
}

/**
 * Method to set active game offline status
 * @param offline boolean
 */
export const setActiveGameOffline = async (offline: boolean) => {
    await AsyncStorage.setItem('active_game_offline', offline.toString())
}

/**
 * Method to set active game id
 * @param id string
 */
export const setActiveGameId = async (id: string) => {
    await AsyncStorage.setItem('active_game_id', id)
}

export const createOfflineGame = async (
    data: CreateGame,
    teamOnePlayers: DisplayUser[],
): Promise<string> => {
    const realm = await getRealm()

    let id: string = ''
    realm.write(() => {
        const game = realm.create<GameSchema>(
            'Game',
            GameSchema.createOfflineGame(data, teamOnePlayers),
        )
        id = game._id
    })

    return id
}

export const saveGame = async (game: Game) => {
    const realm = await getRealm()
    const gameRecord = await realm.objectForPrimaryKey<GameSchema>(
        'Game',
        game._id,
    )
    let offline = false
    if (gameRecord) {
        offline = gameRecord.offline
    }
    const schema = new GameSchema(game, offline)
    realm.write(() => {
        realm.create('Game', schema, Realm.UpdateMode.Modified)
    })
}

export const getGameById = async (
    gameId: string,
): Promise<Game & { offline: boolean }> => {
    const realm = await getRealm()
    const game = await realm.objectForPrimaryKey<GameSchema>('Game', gameId)
    if (!game) {
        return throwApiError({}, Constants.GET_GAME_ERROR)
    }
    return parseGame(game)
}

export const activeGames = async (
    userId: string,
): Promise<(Game & { offline: boolean })[]> => {
    const realm = await getRealm()
    const games = await realm.objects<GameSchema>('Game')

    return games
        .filter(g => g.creator._id === userId)
        .map(game => parseGame(game))
}

export const deleteFullGame = async (gameId: string): Promise<void> => {
    const realm = await getRealm()
    const game = await realm.objectForPrimaryKey<GameSchema>('Game', gameId)

    const points = game?.points.map(id => {
        return realm.objectForPrimaryKey<PointSchema>('Point', id)
    })

    const actions: (ActionSchema | null)[] = []
    points?.forEach(p => {
        p?.teamOneActions.forEach(id => {
            const objectId = new Realm.BSON.ObjectID(id)
            actions.push(
                realm.objectForPrimaryKey<ActionSchema>('Action', objectId),
            )
        })
        p?.teamTwoActions.forEach(id => {
            const objectId = new Realm.BSON.ObjectID(id)
            actions.push(
                realm.objectForPrimaryKey<ActionSchema>('Action', objectId),
            )
        })
    })

    realm.write(() => {
        realm.delete(actions.filter(a => a !== null))
        realm.delete(points?.filter(p => p !== null))
        realm.delete(game)
    })
}
