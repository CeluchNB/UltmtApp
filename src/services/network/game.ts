import { CreateGame } from '../../types/game'
import { GuestUser } from '../../types/user'
import { addQueryParam } from '../../utils/service-utils'
import { API_KEY, API_URL_V1 } from 'react-native-dotenv'
import axios, { AxiosResponse } from 'axios'

export const searchGames = async (
    q?: string,
    live?: boolean,
    after?: string,
    before?: string,
    pageSize?: number,
    offset?: number,
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/game/search`
    url = addQueryParam(url, 'q', q)
    url = addQueryParam(url, 'live', live)
    url = addQueryParam(url, 'after', after)
    url = addQueryParam(url, 'before', before)
    url = addQueryParam(url, 'pageSize', pageSize)
    url = addQueryParam(url, 'offset', offset)
    return await axios.get(url, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const createGame = async (
    token: string,
    data: CreateGame,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/game`,
        { createGameData: data },
        { headers: { 'X-API-Key': API_KEY, Authorization: `Bearer ${token}` } },
    )
}

export const addGuestPlayer = async (
    token: string,
    player: GuestUser,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/game/player/guest`,
        { player },
        { headers: { 'X-API-Key': API_KEY, Authorization: `Bearer ${token}` } },
    )
}

export const getPointsByGame = async (
    gameId: string,
): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/game/${gameId}/points`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const getGameById = async (gameId: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/game/${gameId}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const joinGame = async (
    token: string,
    gameId: string,
    teamId: string,
    code: string,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/game/${gameId}/resolve?team=${teamId}&otp=${code}`,
        {},
        {
            headers: {
                'X-API-Key': API_KEY,
                authorization: `Bearer ${token}`,
            },
        },
    )
}

export const finishGame = async (token: string) => {
    return await axios.put(
        `${API_URL_V1}/game/finish`,
        {},
        {
            headers: {
                'X-API-Key': API_KEY,
                authorization: `Bearer ${token}`,
            },
        },
    )
}

export const getGamesByTeam = async (teamId: string) => {
    return await axios.get(`${API_URL_V1}/game/team/${teamId}`, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}
