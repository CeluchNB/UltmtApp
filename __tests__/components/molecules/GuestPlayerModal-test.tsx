import * as GameServices from '../../../src/services/network/game'
import * as LocalGameServices from '../../../src/services/local/game'
import GuestPlayerModal from '../../../src/components/molecules/GuestPlayerModal'
import { Provider } from 'react-redux'
import React from 'react'
import { game } from '../../../fixtures/data'
import store from '../../../src/store/store'
import { fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

jest.spyOn(LocalGameServices, 'saveGame').mockReturnValue(
    Promise.resolve(undefined),
)

it('should match snapshot', () => {
    const snapshot = render(
        <Provider store={store}>
            <GuestPlayerModal visible={true} onClose={jest.fn()} />
        </Provider>,
    )

    expect(snapshot.toJSON()).toMatchSnapshot()
})

it('should call add player correctly', async () => {
    const spy = jest.spyOn(GameServices, 'addGuestPlayer').mockReturnValue(
        Promise.resolve({
            data: {
                game: { ...game, startTime: '2022', tournament: undefined },
            },
            status: 200,
            statusText: 'Good',
            config: {},
            headers: {},
        }),
    )
    jest.spyOn(LocalGameServices, 'getGameById').mockReturnValueOnce(
        Promise.resolve({
            ...game,
            startTime: '2022' as unknown as Date,
            tournament: undefined,
            offline: false,
        }),
    )
    let visible = true

    const { getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <GuestPlayerModal
                visible={true}
                onClose={() => {
                    visible = false
                }}
            />
        </Provider>,
    )

    const firstInput = getByPlaceholderText('First Name')
    const lastInput = getByPlaceholderText('Last Name')

    fireEvent.changeText(firstInput, 'First')
    fireEvent.changeText(lastInput, 'Last')

    const button = getByText('add')
    fireEvent.press(button)

    await waitFor(() => {
        expect(spy).toHaveBeenCalled()
    })
    expect(visible).toBe(false)
})

it('with form errors', async () => {
    const spy = jest.spyOn(GameServices, 'addGuestPlayer').mockReturnValue(
        Promise.resolve({
            data: {
                game: { ...game, startTime: '2022', tournament: undefined },
            },
            status: 200,
            statusText: 'Good',
            config: {},
            headers: {},
        }),
    )
    spy.mockClear()
    let visible = true

    const { getByText, findByText } = render(
        <Provider store={store}>
            <GuestPlayerModal
                visible={true}
                onClose={() => {
                    visible = false
                }}
            />
        </Provider>,
    )

    const button = getByText('add')
    fireEvent.press(button)

    await findByText('First name is required.')
    expect(spy).not.toHaveBeenCalled()
    expect(visible).toBe(true)
})

it('with network error', async () => {
    const spy = jest.spyOn(GameServices, 'addGuestPlayer').mockReturnValue(
        Promise.reject({
            data: { message: 'Bad guest' },
            status: 400,
            statusText: 'Bad',
            config: {},
            headers: {},
        }),
    )
    spy.mockClear()
    let visible = true

    const { getByPlaceholderText, getByText } = render(
        <Provider store={store}>
            <GuestPlayerModal
                visible={true}
                onClose={() => {
                    visible = false
                }}
            />
        </Provider>,
    )

    const firstInput = getByPlaceholderText('First Name')
    const lastInput = getByPlaceholderText('Last Name')

    fireEvent.changeText(firstInput, 'First')
    fireEvent.changeText(lastInput, 'Last')

    const button = getByText('add')
    fireEvent.press(button)

    await waitFor(() => {
        expect(spy).toHaveBeenCalled()
    })
    expect(getByText('Bad guest')).not.toBeNull()
    expect(visible).toBe(true)
})
