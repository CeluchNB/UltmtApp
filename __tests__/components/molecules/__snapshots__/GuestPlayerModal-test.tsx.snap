// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestPlayerModal should match snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  testID="base-modal"
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#30303050",
        "display": "flex",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#1a1a1a",
          "borderColor": "#ffd600",
          "borderRadius": 10,
          "color": "#ffd600",
          "elevation": 5,
          "flex": 1,
          "margin": 20,
          "padding": 25,
          "shadowColor": "#ffd600",
        }
      }
    >
      <Text
        style={
          {
            "color": "#7AC6FF",
            "fontSize": 41.0958904109589,
          }
        }
      >
        Add Guest Player
      </Text>
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "width": "90%",
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                {
                  "backgroundColor": "#1a1a1a",
                  "borderColor": "#7AC6FF",
                  "borderRadius": 4,
                  "borderWidth": 1,
                },
                undefined,
              ]
            }
            testID="text-input-outline"
          />
          <View>
            <View
              style={
                [
                  {
                    "paddingBottom": 0,
                  },
                  {
                    "minHeight": 56,
                    "paddingTop": 8,
                  },
                ]
              }
            >
              <TextInput
                accessibilityLabel="First Name"
                autoCapitalize="none"
                cursorColor="#ffd600"
                editable={true}
                keyboardType="default"
                maxFontSizeMultiplier={1.5}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="First Name"
                placeholderTextColor="#ffffff"
                selectionColor="#ffd600"
                style={
                  [
                    {
                      "margin": 0,
                      "zIndex": 1,
                    },
                    {
                      "height": 48,
                    },
                    {
                      "paddingBottom": 0,
                      "paddingTop": 0,
                    },
                    {
                      "color": "#ffd600",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": undefined,
                      "letterSpacing": 0.15,
                      "lineHeight": undefined,
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                      "textAlignVertical": "center",
                    },
                    false,
                    [
                      {},
                    ],
                    undefined,
                  ]
                }
                testID="first-name-input"
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "width": "90%",
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              [
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 6,
                },
                {
                  "backgroundColor": "#1a1a1a",
                  "borderColor": "#7AC6FF",
                  "borderRadius": 4,
                  "borderWidth": 1,
                },
                undefined,
              ]
            }
            testID="text-input-outline"
          />
          <View>
            <View
              style={
                [
                  {
                    "paddingBottom": 0,
                  },
                  {
                    "minHeight": 56,
                    "paddingTop": 8,
                  },
                ]
              }
            >
              <TextInput
                accessibilityLabel="Last Name"
                autoCapitalize="none"
                cursorColor="#ffd600"
                editable={true}
                keyboardType="default"
                maxFontSizeMultiplier={1.5}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="Last Name"
                placeholderTextColor="#ffffff"
                selectionColor="#ffd600"
                style={
                  [
                    {
                      "margin": 0,
                      "zIndex": 1,
                    },
                    {
                      "height": 48,
                    },
                    {
                      "paddingBottom": 0,
                      "paddingTop": 0,
                    },
                    {
                      "color": "#ffd600",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "fontWeight": undefined,
                      "letterSpacing": 0.15,
                      "lineHeight": undefined,
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                      "textAlignVertical": "center",
                    },
                    false,
                    [
                      {},
                    ],
                    undefined,
                  ]
                }
                testID="last-name-input"
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": 20,
              "marginLeft": 5,
              "marginRight": 5,
              "marginTop": 10,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "#ffd600",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          {
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "#ffd600",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  cancel
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#ffd600",
              "borderRadius": 20,
              "marginLeft": 5,
              "marginRight": 5,
              "marginTop": 10,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#ffd600",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": "auto",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          {
                            "marginHorizontal": 8,
                          },
                          {
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  add
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
